finality-aleph/src/finalization.rs

```diff
@@ -107,6 +119,185 @@ where
+            let lca = sp_blockchain::lowest_common_ancestor(
+                self.client.as_ref(),
+                reduced,
+                self.last_finalized_hash,
+            )
+            .map_err(|e| error!(target: "afa", "unable to find lca: {:?}", e))
+            .ok()?;
+            if lca.hash != self.last_finalized_hash || reduced == self.last_finalized_hash {
+                error!(target: "afa", "does not extend: last finalized #{:?}, lca #{:?}, reduced #{:?}", self.client.number(self.last_finalized_hash), self.client.number(lca.hash), reduced);
+                continue;
+            }
+            if lca.number == self.max_h {
+                self.finished = true;
+                return None;
+            }
+            self.last_finalized_hash = reduced;
+            return Some(reduced);
+        }
+    }
+}
+
+#[derive(PartialEq, Eq, Hash, Clone, Debug, Encode, Decode)]
+pub(crate) struct SignableHash<H: Codec> {
+    hash: H,
+}
+
+impl<H: AsRef<[u8]> + Hash + Clone + Codec> Signable for SignableHash<H> {
+    type Hash = H;
+    fn hash(&self) -> Self::Hash {
+        self.hash.clone()
+    }
+}
+
+/// A wrapper around an RMC returning the signed hashes in the order of the [`ReliableMulticast::start_rmc`] calls.
+pub(crate) struct BlockSignatureAggregator<'a, B: Block, MK: MultiKeychain> {
+    #[allow(clippy::type_complexity)]
+    messages_for_rmc:
+        mpsc::UnboundedSender<Message<SignableHash<B::Hash>, Signature, SignatureSet<Signature>>>,
+    #[allow(clippy::type_complexity)]
+    messages_from_rmc: mpsc::UnboundedReceiver<(
+        NodeIndex,
+        Message<SignableHash<B::Hash>, Signature, SignatureSet<Signature>>,
+    )>,
+    signatures: HashMap<B::Hash, MK::PartialMultisignature>,
+    hash_queue: VecDeque<B::Hash>,
+    network: RmcNetwork<B>,
+    rmc: ReliableMulticast<'a, SignableHash<B::Hash>, MK>,
+    finished: bool,
+}
+
+impl<
+        'a,
+        B: Block,
+        MK: MultiKeychain<Signature = Signature, PartialMultisignature = SignatureSet<Signature>>,
+    > BlockSignatureAggregator<'a, B, MK>
+{
+    pub(crate) fn new(network: RmcNetwork<B>, keychain: &'a MK) -> Self {
+        let (messages_for_rmc, messages_from_network) = mpsc::unbounded();
+        let (messages_for_network, messages_from_rmc) = mpsc::unbounded();
+        let scheduler = DoublingDelayScheduler::new(Duration::from_millis(10));
+        let rmc = ReliableMulticast::new(
+            messages_from_network,
+            messages_for_network,
+            keychain,
+            keychain.node_count(),
+            scheduler,
+        );
+        BlockSignatureAggregator {
+            messages_for_rmc,
+            messages_from_rmc,
+            signatures: HashMap::new(),
+            hash_queue: VecDeque::new(),
+            network,
+            rmc,
+            finished: false,
+        }
+    }
+
+    pub(crate) async fn start_aggregation(&mut self, hash: B::Hash) {
+        self.hash_queue.push_back(hash);
```

Comment:
Maybe here it would make sense to check if we already received hash?

Reply:
It does :) added a check